tests/herbie.egg

```diff
@@ -0,0 +1,204 @@
+;; Implements the simplification layer of herbie in egg-smol🫡
+
+;; Bools!
+(datatype B (True) (False))
+(function or (B B) B)
+(rewrite (or (True) x) (True))
+(rewrite (or x (True)) (True))
+(rewrite (or (False) (False)) (False))
+
+;; Operators from FPCore
+(datatype Math
+  ; Ground terms
+  (Num Rational)
+  (Var String)
+
+  ; Constant-folding ops
+  (Add Math Math)
+  (Sub Math Math)
+  (Mul Math Math)
+  (Div Math Math)
+  (Pow Math Math)
+  (Neg Math)
+  (Sqrt Math)
+  (Cbrt Math) ; cube root
+  (Fabs Math)
+  (Ceil Math)
+  (Floor Math)
+  (Round Math)
+  (Log Math))
+
+;; This example has four analyses:
+;; an interval analysis consisting of a hi and lo component,
+;; a constant folding analysis, and a nonzero analysis
+(function hi (Math) Rational :merge (min old new))
+(function lo (Math) Rational :merge (max old new))
+(function evals-to (Math) Rational)
+(function is-zero (Math) B
+  :merge (or old new)
+  :default (False))
+
+;; First, define our constant folding analysis
+(rule ((= e (Num n)))
+      ((set (evals-to e) n)))
+(rule ((= e (Add a b))
+       (= va (evals-to a))
+       (= vb (evals-to b)))
+      ((set (evals-to e) (+ va vb))))
+(rule ((= e (Sub a b))
+       (= va (evals-to a))
+       (= vb (evals-to b)))
+      ((set (evals-to e) (- va vb))))
+(rule ((= e (Mul a b))
+       (= va (evals-to a))
+       (= vb (evals-to b)))
+      ((set (evals-to e) (* va vb))))
+(rule ((= e (Div a b))
+       (= va (evals-to a))
+       (= vb (evals-to b))
+       ; TODO: this next one only works if Ratios are reduced in ops
+       ;       should confirm this
+       (!= (rational 0 1) vb))
+      ((set (evals-to e) (/ va vb))))
+; TODO: pow
+(rule ((= e (Neg a))
+       (= va (evals-to a)))
+      ((set (evals-to e) (neg va))))
+; TODO: sqrt
+; TODO: cbrt
+(rule ((= e (Fabs a))
+       (= va (evals-to a)))
+      ((set (evals-to e) (abs va))))
+(rule ((= e (Ceil a))
+       (= va (evals-to a)))
+      ((set (evals-to e) (ceil va))))
+(rule ((= e (Floor a))
+       (= va (evals-to a)))
+      ((set (evals-to e) (floor va))))
+(rule ((= e (Round a))
+       (= va (evals-to a)))
+      ((set (evals-to e) (round va))))
+; TODO: log
+
+;; To check if something is zero, we just check if it evaluates to zero
+(rule ((= (rational 0 1) (evals-to e)))
+      ((set (is-zero e) (True))))
+
+;; The interval analyses are similar to the constant-folding analysis,
+;; except we have to take the lower/upper bound of the results we get
+(rule ((= e (Add a b))
+       (= la (lo a))
+       (= lb (lo b)))
+      ((set (lo e) (+ la lb))))
```

Comment:
In these rules, you could just call `(lo a)` and `(lo b)` in the action. Do you think it's better to do in the query?

Reply:
Hmm, nevermind, that might not work.

Reply:
yeah, if `a` and `b` don't already have a lower bound, this will just panic